web/src/providers/Test/Test.provider.tsx

```diff
@@ -0,0 +1,125 @@
+import useTestCrud from './hooks/useTestCrud';
+import TestService from '../../services/Test.service';
+
+interface IContext {
+  onEdit(values: TDraftTest): void;
+  onRun(): void;
+  isLoading: boolean;
+  isError: boolean;
+  test: TTest;
+  latestTest: TTest;
+  isLatestVersion: boolean;
+  isEditLoading: boolean;
+}
+
+export const Context = createContext<IContext>({
+  onEdit: noop,
+  onRun: noop,
+  test: {} as TTest,
+  latestTest: {} as TTest,
+  isLoading: false,
+  isError: false,
+  isLatestVersion: true,
+  isEditLoading: false,
+});
+
+interface IProps {
+  testId: string;
+  version?: number;
+  children: React.ReactNode;
+}
+
+export const useTest = () => useContext(Context);
+
+const TestProvider = ({children, testId, version = 0}: IProps) => {
+  const [isVersionModalOpen, setIsVersionModalOpen] = useState(false);
+  const [draft, setDraft] = useState<TDraftTest>({});
+  const [state, setState] = useState<'edit' | 'run'>();
+  const {runTest, edit, isEditLoading} = useTestCrud();
+  const {
+    data: test,
+    isLoading: isCurrentLoading,
+    isError: isCurrentError,
+  } = useGetTestVersionByIdQuery({testId, version}, {skip: !version});
+  const {data: latestTest, isLoading: isLatestLoading, isError: isLatestError} = useGetTestByIdQuery({testId});
+
+  const isLatestVersion = useMemo(
+    () => Boolean(version) && version === latestTest?.version,
+    [latestTest?.version, version]
+  );
+  const isLoading = isLatestLoading || isCurrentLoading;
+  const isError = isLatestError || isCurrentError;
+  const currentTest = (version ? test : latestTest)!;
+
+  const onEdit = useCallback(
+    (values: TDraftTest) => {
+      if (isLatestVersion) edit(test!, values);
+      else {
+        setState('edit');
+        setDraft(values);
+        setIsVersionModalOpen(true);
+      }
+    },
+    [edit, isLatestVersion, test]
+  );
+
+  const onRun = useCallback(() => {
+    if (isLatestVersion) runTest(testId);
+    else {
+      setState('run');
+      setIsVersionModalOpen(true);
+    }
+  }, [isLatestVersion, runTest, testId]);
+
+  const onConfirm = useCallback(() => {
+    if (state === 'edit') edit(test!, draft);
+    else {
+      const initialValues = TestService.getInitialValues(test!);
+      edit(test!, initialValues);
+    }
+
+    setIsVersionModalOpen(false);
+  }, [draft, edit, state, test]);
+
+  const value = useMemo<IContext>(
+    () => ({
+      onEdit,
+      onRun,
+      isLoading,
+      isError,
+      test: currentTest,
+      latestTest: latestTest!,
+      isLatestVersion,
+      isEditLoading,
+    }),
+    [onEdit, onRun, isLoading, isError, currentTest, latestTest, isLatestVersion, isEditLoading]
+  );
+
+  return currentTest && latestTest ? (
+    <>
+      <Context.Provider value={value}>{children}</Context.Provider>
+      <VersionMismatchModal
+        description={
+          state === 'edit'
+            ? 'Changing and saving changes will result in a new version that will become the latest.'
```

Comment:
Minor: `Changing and saving changes` sounds weird. Maybe `Updating and saving changes` or just `Editing it will result ...`

Reply:
Updated the copy, maybe we can ask the team for a better one next planning.